### ISSUE TITLE
Discrepancy in Depth Measurements Between Intel RealSense D435i and DA3 Metric Model

### ISSUE THREAD

**Issue by hasarindu-vega** [ByteDance-Seed/Depth-Anything-3#98]
Hi there,

I’m trying to compare the depth metrics using the Intel RealSense D435i and the DA3 Metric Model, but I’m not getting any close values. Am I doing something wrong?

TIA

Depth Comparison Statistics
========================================

Grid: 5x7 (31 valid points)

Mean Absolute Error:     1.8036 m
Root Mean Square Error:  2.0649 m
Mean Relative Error:     56.23 %
Maximum Error:           3.3328 m

RealSense Mean Depth:    2.921 m
DA3 Mean Depth:          1.118 m

<img width="1280" height="480" alt="Image" src="https://github.com/user-attachments/assets/7872ee21-85f6-442e-82be-02e71ecee944" />

### Please find my code below:

```
#!/usr/bin/env python3
"""
RealSense vs DA3 Metric Depth - Grid Comparison

Side-by-side comparison with automatic grid points showing depth values.
Simple and clean visualization with grid overlay.

Requirements:
    pip install pyrealsense2
"""

import cv2
import numpy as np
import torch
from depth_anything_3.api import DepthAnything3
import time
import os
from datetime import datetime

try:
    import pyrealsense2 as rs
    REALSENSE_AVAILABLE = True
except ImportError:
    REALSENSE_AVAILABLE = False
    print("ERROR: pyrealsense2 not installed!")
    print("Install with: pip install pyrealsense2")
    exit(1)

class GridDepthComparison:
    def __init__(self,
                 model_name="depth-anything/DA3METRIC-LARGE",
                 grid_rows=5,
                 grid_cols=7,
                 save_dir="grid_comparison"):
        """
        Initialize comparison system.

        Args:
            model_name: DA3 model to use
            grid_rows: Number of grid rows
            grid_cols: Number of grid columns
            save_dir: Directory to save captures
        """
        self.model_name = model_name
        self.grid_rows = grid_rows
        self.grid_cols = grid_cols
        self.save_dir = save_dir
        os.makedirs(save_dir, exist_ok=True)

        # Initialize systems
        self.init_realsense()
        self.init_da3_model()

        # Display settings
        self.show_grid = True
        self.show_values = True

    def init_realsense(self):
        """Initialize RealSense camera."""
        print("Initializing RealSense camera...")
        self.pipeline = rs.pipeline()
        self.config = rs.config()

        # Configure streams - 640x480 for better performance
        self.config.enable_stream(rs.stream.depth, 640, 480, rs.format.z16, 30)
        self.config.enable_stream(rs.stream.color, 640, 480, rs.format.bgr8, 30)

        # Start streaming
        self.profile = self.pipeline.start(self.config)

        # Get depth scale
        depth_sensor = self.profile.get_device().first_depth_sensor()
        self.depth_scale = depth_sensor.get_depth_scale()

        # Create align object
        self.align = rs.align(rs.stream.color)

        print(f"RealSense initialized - Depth scale: {self.depth_scale}")

    def init_da3_model(self):
        """Initialize DA3 Metric model."""
        print(f"\nLoading DA3 model: {self.model_name}...")
        self.device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
        print(f"Using device: {self.device}")

        self.model = DepthAnything3.from_pretrained(self.model_name)
        self.model = self.model.to(device=self.device)
        self.model.eval()

        print("DA3 model loaded successfully!")

    def get_frames(self):
        """Get aligned frames from RealSense."""
        frames = self.pipeline.wait_for_frames()
        aligned_frames = self.align.process(frames)

        depth_frame = aligned_frames.get_depth_frame()
        color_frame = aligned_frames.get_color_frame()

        if not depth_frame or not color_frame:
            return None, None

        # Convert to numpy arrays
        depth_image = np.asanyarray(depth_frame.get_data())
        color_image = np.asanyarray(color_frame.get_data())

        # Convert depth to meters
        depth_meters = depth_image * self.depth_scale

        return color_image, depth_meters

    def get_da3_depth(self, color_image):
        """Get depth from DA3 model."""
        rgb_image = cv2.cvtColor(color_image, cv2.COLOR_BGR2RGB)

        with torch.no_grad():
            prediction = self.model.inference([rgb_image])

        depth = prediction.depth[0]
        return depth

    def colorize_depth(self, depth, min_val=0.0, max_val=5.0):
        """Convert depth to colored visualization."""
        depth_normalized = (depth - min_val) / (max_val - min_val + 1e-8)
        depth_normalized = np.clip(depth_normalized, 0, 1)
        depth_uint8 = (depth_normalized * 255).astype(np.uint8)
        depth_colored = cv2.applyColorMap(depth_uint8, cv2.COLORMAP_TURBO)
        return depth_colored

    def calculate_grid_points(self, width, height):
        """Calculate grid point positions."""
        # Add margin from edges
        margin_x = width // 10
        margin_y = height // 10

        effective_width = width - 2 * margin_x
        effective_height = height - 2 * margin_y

        step_x = effective_width // (self.grid_cols - 1)
        step_y = effective_height // (self.grid_rows - 1)

        points = []
        for row in range(self.grid_rows):
            for col in range(self.grid_cols):
                x = margin_x + col * step_x
                y = margin_y + row * step_y
                points.append((x, y))

        return points

    def draw_grid_overlay(self, image, depth_map, is_realsense=True, color=(0, 255, 0)):
        """Draw grid with depth values on image."""
        h, w = image.shape[:2]
        overlay = image.copy()

        # Calculate grid points
        grid_points = self.calculate_grid_points(w, h)

        # Draw grid lines if enabled
        if self.show_grid:
            # Horizontal lines
            for row in range(self.grid_rows):
                y = grid_points[row * self.grid_cols][1]
                cv2.line(overlay, (0, y), (w, y), color, 1, cv2.LINE_AA)

            # Vertical lines
            for col in range(self.grid_cols):
                x = grid_points[col][0]
                cv2.line(overlay, (x, 0), (x, h), color, 1, cv2.LINE_AA)

        # Draw points and values
        depths = []
        for x, y in grid_points:
            if 0 <= y < h and 0 <= x < w:
                depth_val = depth_map[y, x]
                depths.append(depth_val)

                # Draw point
                cv2.circle(overlay, (x, y), 4, color, -1)
                cv2.circle(overlay, (x, y), 5, (0, 0, 0), 1)

                # Draw depth value if enabled
                if self.show_values and depth_val > 0:
                    text = f"{depth_val:.2f}"

                    # Text background for readability
                    (text_w, text_h), _ = cv2.getTextSize(text, cv2.FONT_HERSHEY_SIMPLEX, 0.4, 1)
                    cv2.rectangle(overlay,
                                (x - text_w//2 - 2, y - text_h - 8),
                                (x + text_w//2 + 2, y - 5),
                                (0, 0, 0), -1)

                    # Draw text
                    cv2.putText(overlay, text,
                              (x - text_w//2, y - 8),
                              cv2.FONT_HERSHEY_SIMPLEX, 0.4, color, 1, cv2.LINE_AA)

        return overlay, depths

    def calculate_statistics(self, rs_depths, da3_depths):
        """Calculate comparison statistics."""
        rs_array = np.array(rs_depths)
        da3_array = np.array(da3_depths)

        # Filter valid depths
        valid_mask = (rs_array > 0) & (da3_array > 0)

        if not np.any(valid_mask):
            return None

        rs_valid = rs_array[valid_mask]
        da3_valid = da3_array[valid_mask]

        errors = da3_valid - rs_valid
        abs_errors = np.abs(errors)
        rel_errors = (abs_errors / (rs_valid + 1e-8)) * 100

        stats = {
            'count': len(rs_valid),
            'mae': np.mean(abs_errors),
            'rmse': np.sqrt(np.mean(errors**2)),
            'mean_rel_error': np.mean(rel_errors),
            'max_error': np.max(abs_errors),
            'rs_mean': np.mean(rs_valid),
            'da3_mean': np.mean(da3_valid),
        }

        return stats

    def draw_stats_panel(self, width, height, stats, fps, inference_time):
        """Create statistics panel."""
        panel = np.zeros((height, width, 3), dtype=np.uint8)

        y_offset = 40
        x_offset = 20

        # Title
        cv2.putText(panel, "STATISTICS", (x_offset, y_offset),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.7, (255, 255, 0), 2)

        y_offset += 40

        if stats:
            lines = [
                f"Grid Points: {stats['count']}",
                "",
                f"Mean Absolute Error:",
                f"  {stats['mae']:.3f} m",
                "",
                f"Root Mean Square Error:",
                f"  {stats['rmse']:.3f} m",
                "",
                f"Mean Relative Error:",
                f"  {stats['mean_rel_error']:.1f} %",
                "",
                f"Max Error: {stats['max_error']:.3f} m",
                "",
                f"RealSense Mean: {stats['rs_mean']:.2f} m",
                f"DA3 Mean: {stats['da3_mean']:.2f} m",
            ]

            for line in lines:
                cv2.putText(panel, line, (x_offset, y_offset),
                           cv2.FONT_HERSHEY_SIMPLEX, 0.5, (255, 255, 255), 1)
                y_offset += 25

        y_offset += 20
        cv2.putText(panel, f"FPS: {fps:.1f}", (x_offset, y_offset),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 1)
        y_offset += 25
        cv2.putText(panel, f"Inference: {inference_time:.0f}ms", (x_offset, y_offset),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.5, (0, 255, 0), 1)

        # Controls
        y_offset = height - 120
        cv2.putText(panel, "CONTROLS:", (x_offset, y_offset),
                   cv2.FONT_HERSHEY_SIMPLEX, 0.6, (255, 255, 0), 2)
        y_offset += 30

        controls = [
            "G - Toggle grid",
            "V - Toggle values",
            "S - Save",
            "Q - Quit",
        ]

        for ctrl in controls:
            cv2.putText(panel, ctrl, (x_offset, y_offset),
                       cv2.FONT_HERSHEY_SIMPLEX, 0.5, (200, 200, 200), 1)
            y_offset += 25

        return panel

    def save_comparison(self, rs_overlay, da3_overlay, stats):
        """Save comparison data."""
        timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")

        # Save images
        cv2.imwrite(os.path.join(self.save_dir, f"realsense_{timestamp}.png"), rs_overlay)
        cv2.imwrite(os.path.join(self.save_dir, f"da3_{timestamp}.png"), da3_overlay)

        # Save combined view
        combined = np.hstack([rs_overlay, da3_overlay])
        cv2.imwrite(os.path.join(self.save_dir, f"comparison_{timestamp}.png"), combined)

        # Save statistics
        if stats:
            stats_path = os.path.join(self.save_dir, f"stats_{timestamp}.txt")
            with open(stats_path, 'w') as f:
                f.write("Depth Comparison Statistics\n")
                f.write("=" * 40 + "\n\n")
                f.write(f"Grid: {self.grid_rows}x{self.grid_cols} ({stats['count']} valid points)\n\n")
                f.write(f"Mean Absolute Error:     {stats['mae']:.4f} m\n")
                f.write(f"Root Mean Square Error:  {stats['rmse']:.4f} m\n")
                f.write(f"Mean Relative Error:     {stats['mean_rel_error']:.2f} %\n")
                f.write(f"Maximum Error:           {stats['max_error']:.4f} m\n\n")
                f.write(f"RealSense Mean Depth:    {stats['rs_mean']:.3f} m\n")
                f.write(f"DA3 Mean Depth:          {stats['da3_mean']:.3f} m\n")

        print(f"\nSaved comparison to {self.save_dir}/")
        print(f"  Timestamp: {timestamp}")

    def run(self):
        """Main comparison loop."""
        print("\n" + "="*60)
        print("RealSense vs DA3 Metric - Grid Comparison")
        print("="*60)
        print(f"Grid: {self.grid_rows} rows x {self.grid_cols} columns")
        print(f"Total points: {self.grid_rows * self.grid_cols}")
        print("\nControls:")
        print("  G - Toggle grid overlay")
        print("  V - Toggle depth values")
        print("  S - Save comparison")
        print("  Q - Quit")
        print("="*60 + "\n")

        # Warmup
        print("Warming up cameras...")
        for _ in range(30):
            self.pipeline.wait_for_frames()

        print("Starting comparison...\n")

        # FPS tracking
        fps_start = time.time()
        frame_count = 0
        fps = 0
        inference_time = 0

        try:
            while True:
                # Get frames
                color_image, rs_depth = self.get_frames()

                if color_image is None:
                    continue

                # Get DA3 depth
                inf_start = time.time()
                da3_depth = self.get_da3_depth(color_image)
                inference_time = (time.time() - inf_start) * 1000  # ms

                # Resize DA3 depth to match RealSense
                if da3_depth.shape != rs_depth.shape:
                    da3_depth = cv2.resize(da3_depth, (rs_depth.shape[1], rs_depth.shape[0]))

                # Colorize depths
                rs_colored = self.colorize_depth(rs_depth, 0, 5.0)
                da3_colored = self.colorize_depth(da3_depth, 0, 5.0)

                # Draw grid overlays
                rs_overlay, rs_depths = self.draw_grid_overlay(rs_colored, rs_depth,
                                                               is_realsense=True,
                                                               color=(0, 255, 0))
                da3_overlay, da3_depths = self.draw_grid_overlay(da3_colored, da3_depth,
                                                                 is_realsense=False,
                                                                 color=(0, 255, 255))

                # Calculate statistics
                stats = self.calculate_statistics(rs_depths, da3_depths)

                # Calculate FPS
                frame_count += 1
                if time.time() - fps_start >= 1.0:
                    fps = frame_count / (time.time() - fps_start)
                    frame_count = 0
                    fps_start = time.time()

                # Add labels
                h, w = rs_overlay.shape[:2]
                cv2.rectangle(rs_overlay, (0, 0), (w, 35), (0, 0, 0), -1)
                cv2.putText(rs_overlay, "RealSense Depth", (10, 25),
                           cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 0), 2)

                cv2.rectangle(da3_overlay, (0, 0), (w, 35), (0, 0, 0), -1)
                cv2.putText(da3_overlay, "DA3 Metric Depth", (10, 25),
                           cv2.FONT_HERSHEY_SIMPLEX, 0.7, (0, 255, 255), 2)

                # Create stats panel
                stats_panel = self.draw_stats_panel(w//2, h, stats, fps, inference_time)

                # Combine views: [RealSense | DA3 | Stats]
                side_by_side = np.hstack([rs_overlay, da3_overlay])
                display = np.hstack([side_by_side, stats_panel])

                # Show
                cv2.imshow('RealSense vs DA3 Grid Comparison', display)

                # Handle keys
                key = cv2.waitKey(1) & 0xFF

                if key == ord('q'):
                    break
                elif key == ord('s'):
                    self.save_comparison(rs_overlay, da3_overlay, stats)
                elif key == ord('g'):
                    self.show_grid = not self.show_grid
                    print(f"Grid overlay: {'ON' if self.show_grid else 'OFF'}")
                elif key == ord('v'):
                    self.show_values = not self.show_values
                    print(f"Depth values: {'ON' if self.show_values else 'OFF'}")

        except KeyboardInterrupt:
            print("\nInterrupted by user")
        finally:
            self.cleanup()

    def cleanup(self):
        """Cleanup resources."""
        print("\nCleaning up...")
        self.pipeline.stop()
        cv2.destroyAllWindows()
        print("Done!")

def main():
    if not REALSENSE_AVAILABLE:
        return

    try:
        # Configuration
        comparator = GridDepthComparison(
            model_name="depth-anything/DA3METRIC-LARGE",
            grid_rows=5,      # Adjust grid density
            grid_cols=7,      # Adjust grid density
            save_dir="grid_comparison"
        )
        comparator.run()

    except Exception as e:
        print(f"\nError: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()

```

**Comment by Jaykumaran** [comment]
Apply scale factor w.r.t canonical . fx value is normalized to 300, because its common datasets kitti. etc 
Ref: https://github.com/ByteDance-Seed/Depth-Anything-3/issues/19


Check once with the Gradio demo pipeline scripts, its a very good reference which include correct filtering logic as well. 

Also to note that standalone Da3-metric-large metric depth is more accurate than DA3-Nested-Giant-Large though it also used DA3 metric as auxillary model to compute the scale difference


Clone it and run: https://huggingface.co/spaces/depth-anything/depth-anything-3

**Comment by kimmich0615** [comment]
Hi @hasarindu-vega, did your absolute depth estimation improve after using the function apply_metric_scaling_numpy? In my tests, there is still a 1.8x discrepancy in the maximum depth.
